lib/ecto/adapters/sql.ex

```diff
@@ -491,19 +491,22 @@ defmodule Ecto.Adapters.SQL do
   def insert_all(adapter_meta, schema_meta, conn, header, rows, on_conflict, returning, opts) do
     %{source: source, prefix: prefix} = schema_meta
     {_, conflict_params, _} = on_conflict
-    {rows, params} = unzip_inserts(header, rows)
+    {rows, params} = unzip_inserts(header, rows, conn)
     sql = conn.insert(prefix, source, header, rows, on_conflict, returning)
+    # IO.inspect sql |> IO.iodata_to_binary
 
     %{num_rows: num, rows: rows} =
       query!(adapter_meta, sql, Enum.reverse(params) ++ conflict_params, opts)
 
     {num, rows}
   end
 
-  defp unzip_inserts(header, rows) do
+  defp unzip_inserts(header, rows, conn) do
     Enum.map_reduce rows, [], fn fields, params ->
       Enum.map_reduce header, params, fn key, acc ->
         case :lists.keyfind(key, 1, fields) do
+          {^key, {%Ecto.Query{} = query, _maybe_unneeded}} ->
+            {{key, conn.all(query)}, acc}
```

Comment:
I would return the %Ecto.Query{} struct here and let the adapter call `conn.all` :)